File: src/test/resources/files/complex_class/ComplexClass.java
package com.example.test;

public class ComplexClass { // Noncompliant {{Complex class: class complexity 56 is higher than configured: 31.25}}
    public void m1() {
        if (1 > 2) {
            if (2 < 3) {
                while (4 < 3) {
                    for (int i = 0; i < 987; i++) {
                        switch (1) {
                            case 1:
                                if (2 > 4) {
                                    while (3 > 2) {
                                        System.out.println("Hello, world!");
                                    }
                                }
                        }
                    }
                }
            }
        }
    }

    public void m2() {
        if (1 > 2) {
            if (2 < 3) {
                while (4 < 3) {
                    for (int i = 0; i < 987; i++) {
                        switch (1) {
                            case 1:
                                if (2 > 4) {
                                    while (3 > 2) {
                                        System.out.println("Hello, world!");
                                    }
                                }
                        }
                    }
                }
            }
        }
    }
}

File: src/main/scala/io/github/zukkari/checks/ComplexClass.scala
package io.github.zukkari.checks

import io.github.zukkari.base.{ComplexityAccessor, ContextReporter, JavaRule}
import io.github.zukkari.config.ConfigurationProperties
import org.sonar.check.Rule
import org.sonar.plugins.java.api.JavaFileScannerContext
import org.sonar.plugins.java.api.tree.{ClassTree, MethodTree}
import org.sonar.plugins.java.api.tree.Tree.Kind

import scala.jdk.CollectionConverters._

@Rule(key = "ComplexClass")
class ComplexClass
    extends JavaRule
    with ComplexityAccessor
    with ContextReporter {

  private var context: JavaFileScannerContext = _

  private var veryHighClassComplexity: Double = _

  override def scanFile(
      javaFileScannerContext: JavaFileScannerContext): Unit = {
    this.context = javaFileScannerContext

    veryHighClassComplexity = config
      .flatMap(
        _.getDouble(
          ConfigurationProperties.COMPLEX_CLASS_VERY_HIGH_COMPLEXITY.key))
      .orElse(
        ConfigurationProperties.COMPLEX_CLASS_VERY_HIGH_COMPLEXITY.defaultValue.toDouble)

    scan(context.getTree)
  }

  override def scannerContext: JavaFileScannerContext = context

  override def visitClass(tree: ClassTree): Unit = {
    val classComplexity = tree.members.asScala
      .filter(_.is(Kind.METHOD))
      .map(_.asInstanceOf[MethodTree])
      .map(complexity)
      .sum

    report(
      s"Complex class: class complexity $classComplexity is higher than configured: $veryHighClassComplexity",
      tree,
      classComplexity >= veryHighClassComplexity
    )

    super.visitClass(tree)
  }
}
